### ISSUE TITLE
Performance : Druid native ingestion rollup and compaction is ~3x worse as compared EMR based ingestion

### ISSUE THREAD

**Issue by vikramsinghchandel** [apache/druid#10210]
For same Data loads, the druid native ingestion approximately performed 3x worse as compared to EMR-Hadoop based indexing for rollups, and data size.

### 0.18.0 & 0.18.1
Testing was done on both the versions but the performance numbers are provided only for v0.18.1

### Description

**Ingestion Data:** 1-hour data 300 json.gz files, each of ~136Mb in size
**Cluster details:** See Below

**EMR Numbers**: 
|Total Segments|Total Data Size for 17-22-2020 Hour 0|Total Rows|Avg Number of Rows / Segment|Roll Up Ratio|
|-----------------|--------------------------------------|------------|---------------------------------|-------------|
|18|7.2Gb|81M|5M|123|

**Native Ingestion Numbers:** 
|Total Segments|Total Data Size for 17-22-2020 Hour 0|Total Rows|Avg Number of Rows / Segment|Roll Up Ratio|
|----------------|--------------------------------------|-----------|--------------------------------|-------------|
|50|**16.29Gb**|**244M**|4.9M|40|

Please include as much detailed information about the problem as possible.
**Cluster size :** 

- EMR:: 22 * m5.2xl nodes == 176 vCPU

- Native (K8S indexer nodes) = 50 nodes each with 4 CPU and 26G memory (200 vCPU)
-- Each node has 3 workers so a total of 150 workers are spawned for native ingestion.

### Configurations & Spec in use

**EMR-Hadoop Ingestion spec:**

```
{
    "type":"index_hadoop",
    "spec":
    {
      "dataSchema": {
        "dataSource": "<ds name>",
        "parser": {
          "type": "hadoopyString",
          "parseSpec": {
            "format": "json",
            "timestampSpec": {
              "column": "requestTime",
              "format": "millis"
            },
            "dimensionsSpec": {
              "dimensions": [<dimensions>],
              "dimensionExclusions": [],
              "spatialDimensions": []
            }
          }
        },
        "metricsSpec": [<metrics>],
        "granularitySpec": {
          "type": "uniform",
          "segmentGranularity": "HOUR",
          "queryGranularity": "HOUR",
          "intervals" : [ "2020-07-22T00:00:00.000Z/2020-07-22T01:00:00.000Z" ]
        }
      },
      "ioConfig": {
        "type": "hadoop",
        "inputSpec": {
          "type": "static",
          
          "paths": "<path>"
        }
      },
      "tuningConfig" : {
            "type" : "hadoop",
            "partitionsSpec" : {
              "type" : "hashed",
              "partitionDimension" : null,
              "maxRowsPerSegment" : 5000000,
              "assumeGrouped" : false,
              "numShards" : -1
            },
            "shardSpecs" : { },
            "indexSpec" : {
              "bitmap" : {
                "type" : "roaring"
              },
              "dimensionCompression" : "lz4",
              "metricCompression" : "lz4"
            },
            "leaveIntermediate" : false,
            "cleanupOnFailure" : true,
            "overwriteFiles" : true,
            "maxParseExceptions" : 1000,
            "jobProperties" : { },
            "combineText" : false,
            "aggregationBufferRatio" : 0.5,
            "rowFlushBoundary" : 300000,
            "useCombiner" : true,
            "numBackgroundPersistThreads" : 1
      }
    }
}
```
**Native Batch Spec:**

```
{
  "type": "index_parallel",
  "spec": {
    "type": "index_parallel",
    "dataSchema": {
      "dataSource": "<ds name>",
      "timestampSpec": {
        "column": "requestTime",
        "format": "millis"
      },
      "dimensionsSpec": {
        "dimensions": [<dimensions>]
      },
      "metricsSpec": [<metrics>],
      "granularitySpec": {
        "type": "uniform",
        "segmentGranularity": "HOUR",
        "queryGranularity": "HOUR",
        "rollup": true,
        "intervals": [
          "2020-07-21/2020-07-23"
        ]
      }
    },
    "ioConfig": {
      "type": "index_parallel",
      "inputSource": {
        "type": "s3",
        "prefixes": [
          "<path>"
        ]
      },
      "inputFormat": {
        "type": "json"
      }
    },
    "tuningConfig": {
      "type": "index_parallel",
      "partitionsSpec": {
        "type": "hashed",
        "numShards": 50
      },
      "forceGuaranteedRollup": true,
      "totalNumMergeTasks": 100,
      "maxNumSegmentsToMerge": 100,
      "maxNumConcurrentSubTasks": 149,
      "maxRowsInMemory": 4000000,
      "maxPendingPersists": 2,
      "indexSpec": {
        "bitmap": {
          "type": "roaring"
        },
        "dimensionCompression": "lz4",
        "metricCompression": "lz4"
      }
    }
  }
}
```

### Native Indexer runtime confs:

**indexer configs**
druid.worker.version=0
druid.worker.capacity=3
druid.worker.numConcurrentMerges=2

**Peon Processing configs**
druid.processing.numThreads=3
druid.processing.numMergeBuffers=2
druid.peon.defaultSegmentWriteOutMediumFactory.type=offHeapMemory

**TLDR:** Because of the lower roll up ratio and increased number of rows the data size is more than tripled

**Comment by vikramsinghchandel** [comment]
@jihoonson ping

**Comment by jihoonson** [comment]
@vikramsinghchandel thanks for the details! There was a bug in native batch task with rollup (https://github.com/apache/druid/pull/9861) when `partitionDimensions` are not set which was fixed in 0.19. Could you try again but with explicit `partitionDimensions` (you can specify all dimensions in there)? Note that `timestamp` column will not be included in hash partitioning with explicit `partitionDimensions` whereas it will be when they are implicit. Could you compare the rollup ratio of Hadoop and native task with explicit `partitionDimensions`?

Regarding performance, if you still saw difference in performance even with the same rollup ratio, I'm wondering how many actual tasks were running for hadoop and native. Could you double check they used the same number of worker tasks for each map and reduce phases?

BTW, when I tested Indexer last time, it wasn't as good as middleManager in terms of performance because of its limits on the global memory usage and concurrent segment persists/merges. Perhaps you could see different result with middleManagers.

**Comment by vikramsinghchandel** [comment]
> @vikramsinghchandel thanks for the details! There was a bug in native batch task with rollup (#9861) when `partitionDimensions` are not set which was fixed in 0.19. Could you try again but with explicit `partitionDimensions` (you can specify all dimensions in there)? Note that `timestamp` column will not be included in hash partitioning with explicit `partitionDimensions` whereas it will be when they are implicit. Could you compare the rollup ratio of Hadoop and native task with explicit `partitionDimensions`?
> 
> Regarding performance, if you still saw difference in performance even with the same rollup ratio, I'm wondering how many actual tasks were running for hadoop and native. Could you double check they used the same number of worker tasks for each map and reduce phases?
> 
> BTW, when I tested Indexer last time, it wasn't as good as middleManager in terms of performance because of its limits on the global memory usage and concurrent segment persists/merges. Perhaps you could see different result with middleManagers.

Testing this today will share the results.

**Comment by vikramsinghchandel** [comment]
@jihoonson 

Tried the workaround mentioned in #9861 . The rollup and total data size is now comparable to prod(EMR) ingestion. Yes, the native indexer is a bit slower as compared to EMR.

Each MapReduce job utilizes ~500 mappers and ~15 reducers. Each container with 6g mem and 1 vCPU.

Again thanks a lot for pointing out the solution. You can go ahead and closed the ticket. :-)

**Comment by jihoonson** [comment]
Great! Thank you for testing :slightly_smiling_face:

Closing this issue, but please feel free to reopen if you have any further issues.
